src/mapport.cpp

```diff
@@ -0,0 +1,342 @@
+            LogPrintf("natpmp: Port mapping removed successfully.\n");
+        } else {
+            LogPrintf("natpmp: sendnewportmappingrequest(0) failed with %d error.\n", r_send);
+        }
+    }
+
+    closenatpmp(&natpmp);
+    return ret;
+}
+#endif // #ifdef USE_NATPMP
+
+#ifdef USE_UPNP
+static bool ProcessUpnp()
+{
+    bool ret = false;
+    std::string port = strprintf("%u", GetListenPort());
+    const char * multicastif = nullptr;
+    const char * minissdpdpath = nullptr;
+    struct UPNPDev * devlist = nullptr;
+    char lanaddr[64];
+
+    int error = 0;
+#if MINIUPNPC_API_VERSION < 14
+    devlist = upnpDiscover(2000, multicastif, minissdpdpath, 0, 0, &error);
+#else
+    devlist = upnpDiscover(2000, multicastif, minissdpdpath, 0, 0, 2, &error);
+#endif
+
+    struct UPNPUrls urls;
+    struct IGDdatas data;
+    int r;
+
+    r = UPNP_GetValidIGD(devlist, &urls, &data, lanaddr, sizeof(lanaddr));
+    if (r == 1)
+    {
+        if (fDiscover) {
+            char externalIPAddress[40];
+            r = UPNP_GetExternalIPAddress(urls.controlURL, data.first.servicetype, externalIPAddress);
+            if (r != UPNPCOMMAND_SUCCESS) {
+                LogPrintf("UPnP: GetExternalIPAddress() returned %d\n", r);
+            } else {
+                if (externalIPAddress[0]) {
+                    CNetAddr resolved;
+                    if (LookupHost(externalIPAddress, resolved, false)) {
+                        LogPrintf("UPnP: ExternalIPAddress = %s\n", resolved.ToString());
+                        AddLocal(resolved, LOCAL_MAPPED);
+                    }
+                } else {
+                    LogPrintf("UPnP: GetExternalIPAddress failed.\n");
+                }
+            }
+        }
+
+        std::string strDesc = PACKAGE_NAME " " + FormatFullVersion();
+
+        do {
+            r = UPNP_AddPortMapping(urls.controlURL, data.first.servicetype, port.c_str(), port.c_str(), lanaddr, strDesc.c_str(), "TCP", 0, "0");
+
+            if (r != UPNPCOMMAND_SUCCESS) {
+                ret = false;
+                LogPrintf("AddPortMapping(%s, %s, %s) failed with code %d (%s)\n", port, port, lanaddr, r, strupnperror(r));
+                break;
+            } else {
+                ret = true;
+                LogPrintf("UPnP Port Mapping successful.\n");
+            }
+        } while (g_mapport_interrupt.sleep_for(PORT_MAPPING_REANNOUNCE_PERIOD));
+        g_mapport_interrupt.reset();
+
+        r = UPNP_DeletePortMapping(urls.controlURL, data.first.servicetype, port.c_str(), "TCP", 0);
+        LogPrintf("UPNP_DeletePortMapping() returned: %d\n", r);
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        FreeUPNPUrls(&urls);
+    } else {
+        LogPrintf("No valid UPnP IGDs found\n");
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        if (r != 0)
+            FreeUPNPUrls(&urls);
+    }
+
+    return ret;
+}
+#endif // #ifdef USE_UPNP
+
+static void ThreadMapPort()
```

Comment:
I guess at some point in the future we could replace this thread (which does nothing but waiting and periodically run stuff) with scheduler runnables instead? (not in this PR)